replica/compaction_group.hh

```diff
@@ -250,6 +252,9 @@ public:
     storage_group_map& storage_groups() noexcept {
         return _storage_groups;
     }
+    seastar::rwlock& lock() noexcept {
+        return _lock;
+    }
```

Comment:
This (and the storage_group() function) makes it hard to see what's going on. Instead of looking at the class implementation to verify the locking rules, one needs to scan all their dependencies.

It's better to avoid exposing them publicly, but if unavoidable, hide behind helpers: `with_storage_groups_exclusive_lock(std::function<future<> (storage_groups_map&))`.

Reply:
+1 on not exposing storage_groups() publicy anymore, so we know that all callers loop over groups with some lock, even non preemptive ones.

for development flexibility, maybe, we could make a helper that returns a tuple of rwlock guard and storage_group_map&, so we can still do things like:

```
auto [guard, groups] = storage_groups_with_shared_lock();
boost::accumulate(groups, ...).
```

Reply:
we'd probably need a reference wrapper or use a struct, but don't have to be done unless needed.

@Deexie btw, we also have to patch table::parallel_foreach_compaction_group() with shared lock, if not done already.

Reply:
I would prefer to go in the `for_each_storage_group` direction and hide the map from the caller if we don't need access to the map, but rather only the storage groups.  It's the responsibility of the caller to acquire the lock, so the function can accept a reference to the lock holder to force the caller to do so.

Something like `future<> for_each_storage_group(const rwlock::holder&, std::function<future<>(storage_group&)> func)`.

And let's do that in a separate patch, before the dynamic allocation and deallocation of storage groups.